src/orderly/read.py

```diff
@@ -0,0 +1,65 @@
+import ast
+
+
+def orderly_read(path):
+    with open(path) as f:
+        src = ast.parse(f.read())
+    return _read_py(src)
+
+
+# In the R version of this function we do a more involved read, trying
+# to handle "static" versions of all of the orderly core
+# functions. However,we then never actually used that for anything, so
+# to avoid overcomplication we'll do the least possible here which is
+# just to read the parameters from the file, verify that it is only
+# called once, and that is called at the top-level.
+#
+# The return value does nod at future extension though.
+def _read_py(src):
+    ret = {"parameters": []}
+    for expr in src.body:
+        dat = _read_expr(expr)
+        if dat:
+            if dat["name"] == "parameters" and ret["parameters"]:
+                msg = f"Duplicate call to 'parameters()' on line {expr.lineno}"
+                raise Exception(msg)
+            ret["parameters"].append(dat["data"])
+    ret["parameters"] = ret["parameters"][0] if ret["parameters"] else {}
+    return ret
+
+
+def _read_expr(expr):
+    if isinstance(expr, ast.Assign):
+        return _read_expr(expr.value)
```

Comment:
I don't think this really works as expected. It seems like it is trying to support `x = orderly.parameters()` statements. It is going to pass the right-hand side to the recursive `_read_expr()` call, which then matches it against an `ast.Expr` node.

However `Expr` only appears "When an expression, such as a function call, appears as a statement by itself with its return value not used or stored" (https://docs.python.org/3/library/ast.html#ast.Expr). In this case there is no `ast.Expr` node involved, so the check in `_is_orderly_call` will always fail.
```python
>>> print(ast.dump(ast.parse("x = orderly.parameters()"), indent=4))
Module(
    body=[
        Assign(
            targets=[
                Name(id='x', ctx=Store())],
            value=Call(
                func=Attribute(
                    value=Name(id='orderly', ctx=Load()),
                    attr='parameters',
                    ctx=Load()),
                args=[],
                keywords=[]))],
    type_ignores=[])
```

We should probably have separate `_read_stmt` and `_read_expr` functions. The former unwraps `ast.Expr` and `ast.Assign` and passes the inner expression to the latter.

See the grammar for what's an expression and what is a statement: https://docs.python.org/3/library/ast.html#abstract-grammar.

Reply:
yeah, I see I think. I've just set this up to be less clever for now, as nobody is assigning parameter statements I hope